Spanish blueprint doc (#259)
cisco/mindmeld: source/blueprints/screening_app.rst

Health Screening
================

In this step-by-step walkthrough, you'll build a conversational application for a health screening that allows users to identify their risk of prediabetes.

Working through this blueprint will teach you how to

   - configure a MindMeld application in a language different than English
   - handle a large number of domains and intents

>    - learn non-system entities

Reviewer: ```suggestion
   - learn custom entities
```

Since we do not have entity groups in the screening app, we do not need a parser configuration.

9. Using the Question Answerer
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The :doc:`Question Answerer <../userguide/kb>` component in MindMeld is mainly used within dialogue state handlers for retrieving information from the knowledge base. Since the screening blueprint has no knowledge base, question answerer is not needed.

10. Testing and Deployment
^^^^^^^^^^^^^^^^^^^^^^^^^^

Once all the individual pieces (NLP, Dialogue State Handlers) have been trained, configured or implemented, perform an end-to-end test of the app using the :class:`Conversation` class.

.. code:: python

   from mindmeld.components.dialogue import Conversation
   conv = Conversation(nlp=nlp, app_path='screening_app')
   conv.say('quiero conocer mi riesgo')

.. code-block:: console

   ['¿Cuál es su edad?', 'Listening...']

The :meth:`say` method:

 - packages the input text in a user request object

>  - passes the object to the MindMeld Application Manager to a simulate an external user interaction with the app, and

Reviewer: ```suggestion
 - passes the object to the MindMeld Application Manager to simulate an external user interaction with the app, and
```

Health Screening
================

In this step-by-step walkthrough, you'll build a conversational application for a health screening that allows users to identify their risk of prediabetes.

Working through this blueprint will teach you how to

   - configure a MindMeld application in a language different than English
   - learn custom entities
   - use dialogue flows to structure the conversation

.. note::

>    Please make sure to install and run all of the :ref:`pre-requisites <getting_started_virtualenv_setup>` for MindMeld before continuing on with this blueprint tutorial.

Reviewer 2: ```suggestion
   Please make sure to install and run all of the :ref:`pre-requisites <getting_started_virtualenv_setup>` for MindMeld before continuing with this blueprint tutorial.
```

1. The Use Case
^^^^^^^^^^^^^^^

This screening application would be offered by a public health department through a messaging service like SMS or WhatsApp. Users should be able to provide answers to a series of questions asked in Spanish and receive an assessment about their risk for prediabetes. They should be able to opt-out of the questionnaire at any time.
A user's risk is calculated based on the `questionnaire <https://www.cdc.gov/diabetes/risktest/index.html>`_ provided by the American Diabetes Association and the Centers for Disease Control and Prevention.

2. Example Dialogue Interactions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The conversational flows for a health screening will primarily involve asking users a series of questions about their medical history relevant to prediabetes. Once all necessary questions have been asked, the users are informed of their risk for presenting prediabetes.

Reviewer 2: ```suggestion
The conversational flows for a health screening will primarily involve asking users a series of questions about their medical history relevant to prediabetes. Once all necessary questions have been answered, the users are informed of their risk for presenting prediabetes.
```

      ...
      
   @screen_prediabetes.handle(intent='answer_gender')
   def set_gender_send_next(request, responder):
      ...
      
   @screen_prediabetes.handle(intent='answer_yes_gestational')
   def confirm_gestational_send_next(request, responder):
      ...
      
   @screen_prediabetes.handle(intent='answer_yes_family')
   def confirm_family_send_next(request, responder):
      ...
      
   @screen_prediabetes.handle(intent='answer_yes_hbp')
   def confirm_hbp_send_next(request, responder):
      ...
      
   @screen_prediabetes.handle(intent='answer_yes_active')
   def confirm_active_send_next(request, responder):
      ...
      
   @screen_prediabetes.handle(intent='answer_height')
   def set_height_send_next(request, responder):
      ...
      
   @screen_prediabetes.handle(intent='answer_weight')
   def set_weight_send_next(request, responder):
      ...
      
   @app.handle(intent='answer_yes')
   @screen_prediabetes.handle(intent='answer_yes')
   def confirm_send_next(request, responder):
      ...
      
   @app.handle(intent='answer_no')
   @screen_prediabetes.handle(intent='answer_no')
   def negate_send_next(request, responder):
      ...

> The ``@app.dialogue_flow`` decorator designates the flow's entry point. Once inside, every follow up turn will continue to be in this flow until we exit the flow. If the user tries to disrupt the flow by answering something unrelated to the screening, we gently reprompt the user to provide an answer for the current question. We can further control how the screening progresses by setting the ``allowed_intents`` attribute of the params object in the dialogue state handler. This attribute specifies a list of intents that you can set to force the language processor to choose from.

Reviewer 2: ```suggestion
The ``@app.dialogue_flow`` decorator designates the flow's entry point. Once inside, every follow up turn will continue to be in this flow until we exit the flow. If the user response disrupts the flow by answering something unrelated to the prompt, we gently reprompt the user to provide an answer for the current question. We can further control how the screening progresses by setting the ``allowed_intents`` attribute of the params object in the dialogue state handler. This attribute specifies a list of intents that you can set to force the language processor to choose from.
```

Health Screening
================

> In this step-by-step walkthrough, you'll build a conversational application for a health screening that allows users to identify their risk of prediabetes.

Reviewer 3: "...for a health screening that..." --> "...for a health screening application that.."

Working through this blueprint will teach you how to

   - configure a MindMeld application in a language different than English
   - learn custom entities
   - use dialogue flows to structure the conversation

.. note::

   Please make sure to install and run all of the :ref:`pre-requisites <getting_started_virtualenv_setup>` for MindMeld before continuing on with this blueprint tutorial.

1. The Use Case
^^^^^^^^^^^^^^^

This screening application would be offered by a public health department through a messaging service like SMS or WhatsApp. Users should be able to provide answers to a series of questions asked in Spanish and receive an assessment about their risk for prediabetes. They should be able to opt-out of the questionnaire at any time.
A user's risk is calculated based on the `questionnaire <https://www.cdc.gov/diabetes/risktest/index.html>`_ provided by the American Diabetes Association and the Centers for Disease Control and Prevention.

2. Example Dialogue Interactions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The conversational flows for a health screening will primarily involve asking users a series of questions about their medical history relevant to prediabetes. Once all necessary questions have been asked, the users are informed of their risk for presenting prediabetes.

Reviewer 3: The conversational flows for ~~a~~ health screening will primarily ...

   - ``exit`` — Opt-out of the screening and say bye to the user.

The ``prediabetes_screening`` domain supports the following intents:

   - ``opt_in`` — User chooses to undergo the screening.
   - ``answer_age`` — User wants to provide their age.
   - ``answer_gender`` — User wants to provide their gender.
   - ``answer_yes`` — User wants to answer a question with an explicit "yes".
   - ``answer_no`` — User wants to answer a question with an explicit "no".
   - ``answer_yes_gestational`` — User wants to answer the question regarding gestational diabetes with an implied "yes". For example, the answer: "During my first pregnancy" implies an affirmative response.
   - ``answer_yes_family`` — User wants to answer the question regarding family history of diabetes with an implied "yes".
   - ``answer_yes_hbp`` — User wants to answer the question regarding high blood pressure with an implied "yes".
   - ``answer_yes_active`` — User wants to answer the question regarding physical activity with an implied "yes".
   - ``answer_height`` — User wants to provide their height.
   - ``answer_weight`` — User wants to provide their weight.

> There are two types of entities in MindMeld: :ref:`System Entities <system-entities>` and :doc:`Custom Entities <../userguide/entity_recognizer>`. System entities are pre-defined in MindMeld. Examples include ``sys_temperature``, ``sys_time``, and ``sys_interval``. Custom entities are defined by the developers of each application. Within each entity folder, the file ``gazetteer.txt`` contains the full list of values for each custom entity.

Reviewer 3: Should we clarify that the gazetteer offers a strong signal and does not mean we can identify only those entity values?

Change the classification model to random forest instead of the default logistic regression:

.. code:: python

   ic.fit(model_settings={'classifier_type': 'rforest'}, param_selection={'type': 'k-fold', 'k': 10, 'grid': {'class_bias': [0.7, 0.3, 0]}})

.. code-block:: console

    Fitting intent classifier: domain='prediabetes_screening'
    Selecting hyperparameters using k-fold cross-validation with 10 splits
    Best accuracy: 96.90%, params: {'class_weight': {0: 1.0209401709401709, 1: 1.006530612244898, 2: 0.7894047619047618, 3: 1.0645631067961165, 4: 1.268939393939394, 5: 1.07363184079602, 6: 1.4362745098039216, 7: 0.8722477064220183, 8: 0.9802238805970149, 9: 2.202, 10: 1.2960317460317459, 11: 1.1579268292682925}}

You can use similar options to inspect and experiment with the Entity Recognizer and the other NLP classifiers. Finding the optimal machine learning settings is a highly iterative process involving several rounds of model training (with varying configurations), testing, and error analysis. See the :doc:`User Guide <../userguide/nlp>` for more about training, tuning, and evaluating the various MindMeld classifiers.

.. admonition:: Exercise

>    Experiment with different models, features, and hyperparameter selection settings to see how they affect classifier performance. Maintain a held-out validation set to evaluate your trained NLP models and analyze misclassified test instances. Then, use observations from the error analysis to inform your machine learning experimentation. See the :doc:`User Guide <../userguide/nlp>` for examples and discussion.

Reviewer 3: "Maintain a held-out validation set to evaluate your trained NLP models and analyze misclassified ~~test~~ instances."

Just so we are clear that we only analyze errors on the validation set and not the test set.
